### PR TITLE
Add example for 1D simulation of blood flow dynamics

### PR DESCRIPTION
**Pull request by lucapegolotti**
# Modulus Pull Request

## Description
This pull requests adds a new example based on [Learning Reduced-Order Models for Cardiovascular Simulations with Graph Neural Networks.
](https://arxiv.org/pdf/2303.07310.pdf)

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is
up to date with these changes.

### CODE REVIEW

File: examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+""" 
+Download dataset from google drive
+"""
+
+gdown --id 1GGzs3H0EFmhgn5fwwBbrb8l8diRXEMPa
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh:5]: Let us host the dataset on NVIDIA NGC.

Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh:5]: What are the steps to move the data to NVIDIA NGC?

File: examples/cfd/bloodflow_1d_mgn/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+is `training.geometries`, which can take the values `healthy`, `pathological`, 
+`mixed`. These `healthy` and `pathological` refer to the geometries used in
+Section 5.1 and 5.2 of the paper; `mixed` considers all geometries.
+
+## Evaluate Model
+To perform inference on a given model:
+```bash
+python inference.py
+```
+The name of the model needs to be specified in `config.yaml`.
```
[the remark this answers is not in the document]
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/README.md:32]: Can I include a plot from the paper?

File: examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh (hunk 1 above)
Comment (mnabian) [reply examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh:5]: I can work on setting this up

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        params['node_features'] = nodes_features
+        params['edges_features'] = edges_features
+        params['rate_noise'] = cfg.training.rate_noise
+        params['rate_noise_features'] = cfg.training.rate_noise_features
+        params['stride'] = cfg.training.stride
+
+        trainset, testset = train_test_split(graphs, 0.9)
+
+        train_graphs = [graphs[gname] for gname in trainset]
+        traindataset = Bloodflow1DDataset(train_graphs, params, trainset)
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:117]: Can this be changed to `train_dataset` for readability?

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+"""
+    Perform training over all graphs in the dataset.
+
+    Arguments:
+        cfg: Dictionary of parameters.
+
+    """
+if __name__ == "__main__":
+    torch.autograd.set_detect_anomaly(True)
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:304]: Debug code?

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        self.optimizer.zero_grad()
+        loss = 0
+        ns = graph.ndata['next_steps']
+
+        # create mask to weight boundary nodes more in loss
+        mask = torch.ones(ns[:,:,0].shape, device=self.device)
+        imask = graph.ndata['inlet_mask'].bool()
+        outmask = graph.ndata['outlet_mask'].bool()
+
+        bcoeff = 100
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:212]: Can this be specified in the config file?

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                              graph)
+
+            # add prediction by MeshGraphNet to current state
+            new_state = torch.clone(states[-1])
+            new_state[:,0:2] += pred
+            # impose exact flow rate at the inlet (to remove it from loss)
+            new_state[imask,1] = ns[imask,1,istride]
+            states.append(new_state)
+
+            coeff = 0.5
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:237]: Same for this

File: examples/cfd/bloodflow_1d_mgn/README.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+pathological conditions.
+
+<p align="center">
+  <img src="../../../docs/img/bloodflow_1d_mgn_results.gif" alt="Comparison between 
+  the MeshGraphNet prediction and the ground truth for pressure."/>
+</p>
+
+## Dataset
+The dataset is composed of 310 simulations obtained on 8 different
+patient-specific models available in the [Vascular Model Repository](www.vascularmodel.com).
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/README.md:32]: Please add a link to the license file for geometries.

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+import hydra
+
+try:
+    import apex
+except:
+    pass
+
+from modulus.distributed.manager import DistributedManager
+from modulus.models.meshgraphnet import MeshGraphNet
+# from modulus.datapipes.gnn.mgn_dataset import MGNDataset
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:37]: Can we remove this?

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+        params['infeat_nodes'] = infeat_nodes
+        params['infeat_edges'] = infeat_edges
+        params['out_size'] = nout
+        params['node_features'] = nodes_features
+        params['edges_features'] = edges_features
+        params['rate_noise'] = cfg.training.rate_noise
+        params['rate_noise_features'] = cfg.training.rate_noise_features
+        params['stride'] = cfg.training.stride
+
+        trainset, testset = train_test_split(graphs, 0.9)
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:114]: Probably better to specify the train/test split in the config file

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+from generate_dataset import train_test_split
+from generate_dataset import Bloodflow1DDataset
+
+from modulus.launch.logging import (
+    PythonLogger,
+    initialize_wandb,
+    RankZeroLoggingWrapper,
+)
+from modulus.launch.utils import load_checkpoint, save_checkpoint
+import argparse
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:49]: `argparse` is not used, please remove it

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+        cfg: Dictionary of parameters.
+
+    """
+    logger = PythonLogger("main")
+    logger.file_logging()
+    trainer = MGNTrainer(logger, cfg)
+    start = time.time()
+    logger.info("Training started...")
+    for epoch in range(trainer.epoch_init, cfg.training.epochs):
+        for graph in trainer.dataloader:
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/train.py:262]: I'm just curious, why did you choose not to shuffle the different time steps? Currently for a given case, the training is performed over multiple timesteps sequentially. But you could sample any case at any timestep instead, similar to the vortex shedding example.

File: examples/cfd/bloodflow_1d_mgn/generate_graphs.py [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+
+    times = [t for t in field]
+    times.sort()
+    times = times[offset:]
+
+    loading_t = th.zeros((list(field.values())[0].shape[0], 1,
+                          len(timesteps) - offset + pad), dtype = th.bool)
+
+    if pad > 0:
+        # def interpolate_function(count):
```
Comment (mnabian) [inline examples/cfd/bloodflow_1d_mgn/generate_graphs.py:70]: Do we need to keep this?

**Comment by mnabian** [comment]
@lucapegolotti the formatting test is failing. Please install `black` and run `black .` in the `modulus-launch` directory. Also if you are using vscode, please install `markdownlint` and fix the issues.

File: examples/cfd/bloodflow_1d_mgn/graph_tools.py [hunk 13]
```diff
@@ -0,0 +1,543 @@
...
+    sampled_indices, edges1, edges2 = remove_points(ipoints_to_delete,
+                                                    ipoints_to_replace,
+                                                    edges1, edges2,
+                                                    npoints)
+
+    points = np.delete(points, ipoints_to_delete, axis = 0)
+
+    return sampled_indices, points, edges1, edges2, indices
+
+def dijkstra_algorithm(nodes, edges1, edges2, index):
```
Comment (akshaysubr) [inline examples/cfd/bloodflow_1d_mgn/graph_tools.py:243]: Would it make sense to move some of these more general utilities into modulus-core?

File: examples/cfd/bloodflow_1d_mgn/inference.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+from modulus.launch.utils import load_checkpoint
+import hydra
+from omegaconf import DictConfig, OmegaConf
+import json
+import time
+
+def denormalize(tensor, mean, stdv):
+        return tensor * stdv + mean 
+
+class MGNRollout:
```
Comment (akshaysubr) [inline examples/cfd/bloodflow_1d_mgn/inference.py:43]: docstrings for this class would be good to add

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+    Returns:
+        The mean square error
+
+    """
+    return (mask * (input - target) ** 2).mean() 
+
+class MGNTrainer:
+    def __init__(self, logger, cfg):
+        # set device
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (akshaysubr) [inline examples/cfd/bloodflow_1d_mgn/train.py:74]: This device should be set to `DistributedManager().device`

File: examples/cfd/bloodflow_1d_mgn/train.py [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+                epoch=epoch,
+        )
+        start = time.time()
+        trainer.scheduler.step()
+
+        def default(obj):
+            if isinstance(obj, torch.Tensor):
+                return default(obj.detach().numpy())
+            if isinstance(obj, np.ndarray):
+                return obj.tolist()
```
Comment (akshaysubr) [inline examples/cfd/bloodflow_1d_mgn/train.py:286]: Wouldn't it be better to do `obj.tobytes()` here to avoid saving the whole tensor/array as ascii?

File: examples/cfd/bloodflow_1d_mgn/generate_dataset.py [hunk 17]
```diff
@@ -0,0 +1,671 @@
+# Copyright 2023 Stanford University
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
```
Comment (akshaysubr) [inline examples/cfd/bloodflow_1d_mgn/generate_dataset.py:3]: Is there a specific reason you chose this license text rather than Apache 2.0?

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 15 above)
Comment (mnabian) [reply examples/cfd/bloodflow_1d_mgn/train.py:77]: @akshaysubr distributed training is not supported in this PR. We will add DDP support in a subsequent PR.

File: examples/cfd/bloodflow_1d_mgn/generate_graphs.py (hunk 12 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/generate_graphs.py:73]: Deleted the commented part

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 8 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/train.py:38]: Done

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 4 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/train.py:303]: Sorry I added this to debug but I will remove it.

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 15 above)
Comment (akshaysubr) [reply examples/cfd/bloodflow_1d_mgn/train.py:77]: I got that, but would still be good to standardize on having all device logic in the `DistributedManager`. So things like the issue raised in https://github.com/NVIDIA/modulus/issues/115 can be handled in a unified manner.

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 11 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/train.py:261]: I am passing the shuffle argument to the GraphDataLoader and just checked by printing graph stats that indeed graphs are passed in a random order at each epoch. Is this what you were confused about?

Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/train.py:261]: See https://github.com/NVIDIA/modulus-launch/blob/de467f3d386ca9912460ea2b182584f8802e8f44/examples/cfd/bloodflow_1d_mgn/train.py#L123

File: examples/cfd/bloodflow_1d_mgn/generate_dataset.py (hunk 17 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/generate_dataset.py]: I just asked my advisor, and she advised me to use MIT, but if you have a preference for Apache 2.0 I could ask whether that would be acceptable. Would it be better for consistency?

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 9 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/train.py]: Done

File: examples/cfd/bloodflow_1d_mgn/README.md (hunk 7 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/README.md:32]: Done

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 16 above)
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/train.py:284]: I will simplify this. The tensors are only scalars in this case.

File: examples/cfd/bloodflow_1d_mgn/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucapegolotti) [reply examples/cfd/bloodflow_1d_mgn/README.md]: I think I'll keep it this way. The paper is currently under review and I am not sure if I can post the pictures from it.

File: examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh (hunk 1 above)
Comment (mnabian) [reply examples/cfd/bloodflow_1d_mgn/raw_dataset/download_dataset.sh:20]: Let us keep it as is, and I'll update the dataset download instructions once the NGC dataset is set up.

File: examples/cfd/bloodflow_1d_mgn/inference.py (hunk 14 above)
Comment (mnabian) [reply examples/cfd/bloodflow_1d_mgn/inference.py:48]: Added

File: examples/cfd/bloodflow_1d_mgn/train.py (hunk 15 above)
Comment (mnabian) [reply examples/cfd/bloodflow_1d_mgn/train.py]: Fixed this

File: examples/cfd/bloodflow_1d_mgn/generate_dataset.py (hunk 17 above)
Comment (mnabian) [reply examples/cfd/bloodflow_1d_mgn/generate_dataset.py]: Updated the license headers to Apache 2.0

**Review by mnabian — APPROVED** [review]
